[docs] EKS & AKS auto-discovery docs (#17628)
gravitational/teleport: docs/pages/kubernetes-access/guides/discovery/aws.mdx

---

> title: Teleport EKS auto-discovery (Preview)

Reviewer: ```suggestion
title: Teleport EKS Auto-Discovery (Preview)
```

description: Auto-discovery of EKS clusters in AWS cloud.
---

> Similarly to [AKS discovery](./azure.mdx), the EKS Discovery can automatically 

Reviewer: ```suggestion
Similarly to [AKS discovery](./azure.mdx), EKS Auto-Discovery can automatically 
```

> discover any EKS cluster running and enroll it if its tags match the configured labels. 

Reviewer: ```suggestion
discover any EKS cluster running and enroll it in Teleport if its tags match the configured labels. 
```

> As mentioned in [Kubernetes Discovery](../discovery.mdx), the discovery process 

Reviewer: As with an earlier comment, I think we should use "Kubernetes Service" instead of `kubernetes_service` and "Discovery Service" instead of `discovery_service` when we are talking about the service instead of its configuration block.

requires two separate Teleport components. The first, `discovery_service`, is 
responsible for watching the cloud provider and checking if there are any 
new clusters or if there have been any modifications to previously discovered clusters. 
The `kubernetes_service` monitors the clusters created by `discovery_service` and 
is responsible for authenticating into them and forwarding the requests.

<Notice type="tip">

This guide presents the `discovery_service` and `kubernetes_service` running in 
the same process, however both can run independently and on different machines. To do 
that, you just need to create a valid token for each of the roles and separate their 
configurations. Running the processes independently is beneficial if access 
to different clusters requires `kubernetes_service` to be on the same network as 
the cluster.

</Notice>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

> - AWS IAM permissions to list and describe EKS clusters and IAM access to the cluster.

Reviewer: Your AWS user should have these permissions?

- EKS clusters running. 

> ## Known Limitations

Reviewer: ```suggestion
## Known limitations
```

Due to the authentication mode that AWS has chosen for EKS clusters, it is not possible 

> for the Teleport process to set up access for itself without previously having access to the cluster.

Reviewer: Does "the cluster" mean "an EKS cluster Teleport is attempting to register"?

This is because the authentication method requires the Teleport IAM Role to be configured 
in a `configmap/aws-auth` that exists in the `kube-system` namespace.

> Please ensure that the permissions are set correctly if the Teleport process does not use 

Reviewer: Are there some handy commands that an operator can copy/paste to determine this?

Author: Yes but nothing simple.

To test that they can:
- Inspect the `aws-auth` config and validate if the Teleport IAM role is there.
- Check if ClusterRole and ClusterRoleBinding exist.

Or:
- Manually create a Token for the IAM role they want to use in Teleport
- Setup a kubeconfig for the cluster
- use `kubectl can-i ...` to check if everything works

the same role as the cluster creator.

## Step 1/3. Set up AWS IAM credentials

For Teleport to discover EKS clusters, the instance
performing discovery requires an IAM policy that allows Teleport to list and 
describe EKS clusters:

```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "eks:DescribeCluster",
                "eks:ListClusters"
            ],
            "Resource": "*"
        }
    ]
}      
```

## Step 2/3. Configure EKS authentication

> When `kubernetes_service` uses an IAM role that is different from the one that 

Reviewer: > When `kubernetes_service` uses an IAM role that is different from the one that created the cluster, 

Should we expect this to be the case for most users completing this guide?

Author: Yes.

created the cluster, it is necessary to edit the `configmap/aws-auth` 
and add the mapping between the IAM role that the Teleport process is using and the 
RBAC permissions for each of the desired clusters. You can take advantage of EKS 
cluster creation and management tools such as `eksctl` and Terraform to automatically 
configure access.

For simplicity, it is possible to set up an existing group in the cluster to the 
Teleport IAM role such as `system:masters`, but recommend setting up the minimum 
permissions as described below.

> <Details 

Reviewer: If this is part of the steps we expect users to complete while following this guide, I think we should move it outside of a `Details` box and make it part of the main body text in Step 2. Otherwise, it's easy to miss.

- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
- apiGroups:
  - "authorization.k8s.io"
  resources:
  - selfsubjectaccessreviews
  - selfsubjectrulesreviews
  verbs:
  - create
```

### ClusterRoleBinding

Link the previously created `ClusterRole` into an RBAC group.

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: teleport
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: teleport
subjects:
- kind: Group
  name: teleport
  apiGroup: rbac.authorization.k8s.io
```

Finally, edit the `configmap/aws-auth` in the `kube-system` namespace to append 
the following `mapRoles`. Please replace `{teleport_aws_iam_role}` with the 
appropriate IAM role. This step will link the Teleport IAM role into the Kubernetes 
RBAC group `teleport`.

```yaml
apiVersion: v1
data:
  mapRoles: |
    - groups:
      - teleport
      rolearn: {teleport_aws_iam_role} # e.g. arn:aws:iam::222222222222:role/teleport-role
      username: teleport
```

At this point, the Teleport IAM role already has the minimum permissions 
to forward requests to the cluster. 

</Details>

## Step 3/3. Configure Teleport to discover EKS clusters

> The Teleport Kubernetes Discovery requires a valid auth token for the `discovery` and `kube` services 

Reviewer: ```suggestion
Teleport EKS Auto-Discovery requires a valid auth token for the `discovery` and `kube` services 
```
To be consistent with other parts of the guide (though are we capitalizing "Auto-Discovery" everywhere?).

to connect to the cluster. Generate one by running the following command against
your Teleport Auth Service and save it in `/tmp/token` on the Node

> that will run the Kubernetes Discovery:

Reviewer: ```suggestion
that will run Kubernetes Auto-Discovery:
```

```code
$ tctl tokens add --type=discovery,kube
```

> To enable EKS cluster discovery, the `discovery_service.aws` section must

Reviewer: > EKS cluster discovery

Let's use the same term throughout the page

include at least one entry and must specify the `eks` type. It 
is also required to configure the `kubernetes_service.resources.tags` to use the same
labels configured at `discovery_service.aws[:].tags` or a subset of them 
in order to make `kubernetes_service` listen to the dynamic resources created by 

> the discovery service.

Reviewer: ```suggestion
the Discovery Service.
```
